### ISSUE TITLE
Avatars

### ISSUE THREAD

**Issue by funkaoshi** [sdelements/lets-chat#4]
We should have pretty circular avatars ala Demiba or Path.

**Comment by jiyinyiyong** [comment]
Bad news, gravatar is not available in mainland China.
![empty-page](https://cloud.githubusercontent.com/assets/449224/6211228/eb7a93b6-b612-11e4-910c-4f5aa42fabca.png)

**Comment by TuxOtaku** [comment]
Yeah I've been thinking for a while that we should support another method for profile pics aside from Gravatar.

**Comment by hhaidar** [comment]
Re-opening this.

@jiyinyiyong do you know of any alternatives that we can fallback too?

**Comment by jiyinyiyong** [comment]
@hhaidar I think encouraging users to upload images is my best solution(cropping is not neccessary though).

**Comment by hhaidar** [comment]
@jiyinyiyong do you think this is something you can work on? It might be a while before we get to it as our team has access to Gravatar.

**Comment by sibartlett** [comment]
Avatar upload functionality should use the upload plugin currently in use (local/S3).

**Comment by jiyinyiyong** [comment]
@hhaidar I'm not quite familiar with server side programming and was not planning to do that :(

**Comment by sibartlett** [comment]
@jiyinyiyong no worries :)

**Comment by bnvk** [comment]
There is [libravatar.org](https://www.libravatar.org/) which is a more open Gravatar

You could also make a tool that allows ppl to input their Twitter @handle or Facebook username and then it goes and pulls in that avatar!

**Comment by pathetiq** [comment]
+1 for twitter photo or even a local directory where we can upload each user photo. each photo can be link to a user id. This way it can be possible to add a local/gravatar/other sections in the settings.yml.

**Comment by sibartlett** [comment]
We're planning to use Mongo's binary field support for avatar uploads - so we don't need to rely on a file system (local/S3/etc).

**Comment by hhaidar** [comment]
A little preview of what's to come: ![screen shot 2015-06-22 at 4 43 59 pm](https://cloud.githubusercontent.com/assets/1066160/8292560/528013ee-18fe-11e5-8f42-971fab45ea84.png)

**Comment by 8bitjeremy** [comment]
Sweet!

**Comment by superlou** [comment]
Is there a pull-request or branch available to try this feature?

**Comment by hhaidar** [comment]
@superlou ya I have a really barebones implementation in #521 

You can add something like this to your settings.yml now:

```
avatars:
  providers: [gravatar, initials]
```

It'll try gravatar first and then go to an SVG generator. I recommend removing gravatar if you're not going to use it.

**Comment by legel** [comment]
Hey @hhaidar 

I was just wondering what it would take to get this feature finished up.  The screenshot looks really nice and in general the initials are a cleaner, smarter fallback than the same generic logo Gravitar assigns to everyone... which gives rise to a lot of visual debt in the application, as more and more users who aren't on Gravitar sign up.

Thanks in any case for the excellent work, let me know if there's something specifically I can help with.

Lance

**Comment by craSH** [comment]
If using LDAP for user authentication, it could be nice to pull user photos from the directory as well. 

Currently gravatar.com is used for profile photos. Consider first trying to use profile photos for the user from their LDAP entry. The available, standard fields are `jpegPhoto` (preferred, can be high-res/no specific size limit) and `thumbnailPhoto` (limited to a fairly small size).

- jpegPhoto: https://msdn.microsoft.com/en-us/library/cc221100.aspx
- thumbnailPhoto: https://msdn.microsoft.com/en-us/library/cc221395.aspx

It would be nice to be able to disable Gravatar in the case of using AD/LDAP auth as well.

Additionally, the login page will render a Gravatar for any email account with a Gravatar - including those not permitted to login to the system. There should be a check to prevent outside users' gravatars from loading at all, it may lead users to believe they have an account on the system when they do not.

**Comment by OliPelz** [comment]
Functionality of uploading local image files instead of relying on an external image service really is important for our department (governmental) because we need a chat application which is 100% self-hosted with zero connections to another external service. At the current state we cannot use it because this is a no-go for us! 
Please add this feature!
